### PR TITLE
Conversion to Core; plumbing

### PR DESCRIPTION
**Pull request by amosr**
Some work on converting Icicle to Core, and something to make the parse -> typecheck -> convert process easier

### CODE REVIEW

**Comment by amosr** [comment]
We can convert a few trivial queries from Icicle to Core now!

```
>> mapRight pretty $ sourceParseConvert "feature salary ~> filter value > 0 ~> sum (value * value) / (sum value * sum value)"
Right Program (source : Stream Int)
Precomputations:

Stream transformers:
v0 =
    source
v1 =
    sfilter [Int] (gt# [Int] v2 (0 : Int)) v0
Reductions:
v4 =
    rfold   [Int] [Int] (\v5 : Int. \v3 : Int. add# v5 (mul# v3 v3)) (0 : Int) v1
v7 =
    rfold   [Int] [Int] (\v8 : Int. \v6 : Int. add# v8 v6) (0 : Int) v1
v10 =
    rfold   [Int] [Int] (\v11 : Int. \v9 : Int. add# v11 v9) (0 : Int) v1
Postcomputations:
v12 =
    mul# v7 v10
v13 =
    div# v4 v12
Returning:
    v13
```

File: src/Icicle/Repl.hs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+import qualified        Icicle.Source.Query             as SQ
+import qualified        Icicle.Source.ToCore.ToCore     as STC
+import qualified        Icicle.Source.ToCore.Base       as STC
+import qualified        Icicle.Source.Type              as ST
+
+import                  P
+import                  Data.Either.Combinators
+
+import qualified        Data.Map as Map
+import qualified        Data.Text as T
```
Comment (tmcgilchrist) [inline src/Icicle/Repl.hs]: Formatting and alphabetisation of imports /nitpick

Comment (amosr) [inline src/Icicle/Repl.hs]: didn't you know "D" was after "P"?

Comment (tmcgilchrist) [inline src/Icicle/Repl.hs]: C-c C-. will fix all those issues. ;-)

**Comment by amosr** [comment]
I am merging this without emoticons so @tranma can do stuff

**Comment by markhibberd** [comment]
:snowman:

**Pull request merged**